File: 11월 1주차/백준 1717 그래프 집합의 표현.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> parent;     // 전역변수 parent

int find(int x) {
    if (parent[x] != x) {
        parent[x] = find(parent[x]);
    }
    return parent[x];
}

void unionSet(int x, int y) {
    x = find(x);
    y = find(y);
    if (x != y) {
        parent[y] = x;
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;
    parent.resize(n + 1);
    for (int i = 0; i <= n; i++) {
        parent[i] = i;
    }
    for (int i = 0; i < m; i++) {
        int op, a, b;
        cin >> op >> a >> b;
        if (op == 0) {
            unionSet(a, b);
        } else if (op == 1) {
            if (find(a) == find(b)) {
                cout << "YES\n";
            } else {
                cout << "NO\n";
            }
        }
    }
    return 0;
}

File: 11월 1주차/백준 1717 그래프 집합의 표현.py
import sys
input = sys.stdin.readline

# 루트 노드가 아니라면 루트 노드를 찾을 때까지 재귀적으로 호출
def find(x):
    if parent[x] != x:
        parent[x] = find(parent[x])
    return parent[x]

# 두 원소가 속한 집합을 합치기
def union(x, y):
    x = find(x)
    y = find(y)
    
    if x != y:
        parent[y] = x

n, m = map(int, input().split())
parent = [i for i in range(n+1)]

for _ in range(m):
    op, a, b = map(int, input().split())
    
    if op == 0:
        union(a, b)
    elif op == 1:
        if find(a) == find(b):
            print("YES")
        else:
            print("NO")
